syntax = "proto3";

package arduino_trader.planning;

import "common/common.proto";
import "common/position.proto";
import "common/security.proto";

option go_package = "github.com/arduino-trader/contracts/planning";

// Planning Service - Handles portfolio planning
service PlanningService {
  // Create a new portfolio plan (streaming for progress updates)
  rpc CreatePlan(CreatePlanRequest) returns (stream PlanUpdate);

  // Get an existing plan
  rpc GetPlan(GetPlanRequest) returns (GetPlanResponse);

  // List all plans for a portfolio
  rpc ListPlans(ListPlansRequest) returns (ListPlansResponse);

  // Get best result for a portfolio
  rpc GetBestResult(GetBestResultRequest) returns (GetBestResultResponse);

  // Health check
  rpc HealthCheck(common.Empty) returns (common.HealthCheckResponse);
}

// Request to create a new plan
message CreatePlanRequest {
  string portfolio_hash = 1;
  repeated common.Position positions = 2;
  common.Money available_cash = 3;
  common.Money monthly_deposit = 4;
  int32 max_positions = 5;
  double target_allocation_pct = 6;
  map<string, string> constraints = 7;
}

// Progress update during planning
message PlanUpdate {
  string plan_id = 1;
  int32 progress_pct = 2;
  string current_step = 3;
  bool complete = 4;
  Plan plan = 5;  // Only populated when complete=true
  string error = 6;  // Only populated if error occurred
}

// Plan represents a portfolio plan
message Plan {
  string id = 1;
  string portfolio_hash = 2;
  repeated PlannedAction actions = 3;
  double score = 4;
  common.Money expected_cost = 5;
  common.Money expected_value = 6;
  common.Timestamp created_at = 7;
  PlanStatus status = 8;
}

// Planned action (buy/sell)
message PlannedAction {
  common.TradeSide side = 1;
  string symbol = 2;
  string isin = 3;
  double quantity = 4;
  common.Money estimated_price = 5;
  common.Money estimated_cost = 6;
  string reason = 7;
  int32 priority = 8;
}

// Plan status
enum PlanStatus {
  PLAN_STATUS_UNSPECIFIED = 0;
  DRAFT = 1;
  READY = 2;
  EXECUTING = 3;
  COMPLETED = 4;
  CANCELLED = 5;
}

// Get plan request
message GetPlanRequest {
  string plan_id = 1;
}

// Get plan response
message GetPlanResponse {
  bool found = 1;
  Plan plan = 2;
}

// List plans request
message ListPlansRequest {
  string portfolio_hash = 1;
  int32 limit = 2;
  int32 offset = 3;
}

// List plans response
message ListPlansResponse {
  repeated Plan plans = 1;
  int32 total = 2;
}

// Get best result request
message GetBestResultRequest {
  string portfolio_hash = 1;
}

// Get best result response
message GetBestResultResponse {
  bool found = 1;
  Plan plan = 2;
}
